Generate should skip unnecessary work

nickpape-msft opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/microsoft/web-build-tools/133)
<!-- Reviewable:end -->

nickpape-msft commented:
Reviewed 1 of 7 files at r1.
Review status: 1 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[rush/rush/src/actions/GenerateAction.ts, line 27 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEMBbFkfgx5rYS8NJ1:-KfEMBbFkfgx5rYS8NJ2:b-q5219y) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L27)):*
> ```XML
> import { TempModuleGenerator } from '../utilities/TempModuleGenerator';
> 
> interface IShrinkwrapFile {
> ```

I have a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Review status: 1 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[rush/rush/src/actions/GenerateAction.ts, line 27 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEMBbFkfgx5rYS8NJ1:-KfEMsA9e2uc_ybiT_Y0:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L27)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

I have a comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

iclanton commented:
Review status: 0 of 7 files reviewed at latest revision, 14 unresolved discussions.

---

*[rush/rush/src/actions/GenerateAction.ts, line 186 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEORoDei151Cx4LpFp:-KfEORoDei151Cx4LpFq:b-ivggkm) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L186)):*
> ```XML
> 
>     const shrinkwrapFile: string = path.join(rushConfiguration.commonFolder, 'npm-shrinkwrap.json');
>     if (!fsx.existsSync(shrinkwrapFile)) {
> ```

Instead of hitting the disk twice, log and return `true` if `require` below throws.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfEONlBcrBKhuG5x6-E:bm022b5) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
> ```XML
>     }
> 
>     const shrinkwrap: IShrinkwrapFile = JSON.parse( fsx.readFileSync(shrinkwrapFile).toString() );
> ```

Use `require`.

---

*[rush/rush/src/actions/GenerateAction.ts, line 199 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfENpXxPuM9MCcIB1bC:-KfENpXxPuM9MCcIB1bD:bfks1x7) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L199)):*
> ```XML
>       const tempProjectName: string = rushConfiguration.projectsByName.get(projectName).tempProjectName;
>       Object.keys(project.dependencies).forEach((dependency: string) => {
>         // technically we need to look at the temp_modules dependencies
> ```

What do you mean "technically?

---

*[rush/rush/src/actions/GenerateAction.ts, line 203 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOJDrdvOpACdW-Ai0:-KfEOJDrdvOpACdW-Ai1:bcyf7h5) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L203)):*
> ```XML
>         if (!GenerateAction._canFindDependencyInShrinkwrap(shrinkwrap, dependency, version, tempProjectName)) {
>           console.log(colors.yellow(
>             `${os.EOL}Could not find "${projectName}" dependency "${dependency}@${version}" in shrinkwrap.`));
> ```

Why are you prepending with `EOL`?

---

*[rush/rush/src/actions/GenerateAction.ts, line 208 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOdikIHk0Zu7IeoMp:-KfEOdil65QooW4OETDD:b88a930) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L208)):*
> ```XML
>       });
>     });
>     if (isLazy) {
> ```

Does lazy mode really serve much purpose anymore? It skips the fastest step.

---

*[rush/rush/src/actions/GenerateAction.ts, line 213 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOjh71K3j1ercG0FN:-KfEOjh8t50v_raQlqex:b-6agyeh) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L213)):*
> ```XML
>     } else if (!hasFoundMissingDependency) {
>       console.log(colors.green(
>         `${os.EOL}Rush found all dependencies in the shrinkwrap! Rush now running in "fast" mode.`));
> ```

I don't think you need to say it's changing "modes."

---

*[rush/rush/src/actions/GenerateAction.ts, line 225 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEPic7ptAfyojpZnzb:-KfEPic7ptAfyojpZnzc:b-64dx74) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L225)):*
> ```XML
>     dependency: string,
>     version: string,
>     rushPackageName: string): boolean {
> ```

You call this `tempProjectName` above. Can you call them the same thing?

---

*[rush/rush/src/actions/GenerateAction.ts, line 230 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEPv3Podsh6U0-w49a:-KfEPv3Podsh6U0-w49b:b-2oxhbv) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L230)):*
> ```XML
> 
>     // The dependency will either be directly under the rushPackageName, or it will be in the root
>     if (shrinkwrap.dependencies[rushPackageName]) {
> ```

I'd look in the root of the shrinkwrap first because that's where we prefer dependencies to live.

---

*[rush/rush/src/actions/GenerateAction.ts, line 238 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ21tgiN_srK0rF7F:-KfEQ21tgiN_srK0rF7G:b-uszk1s) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L238)):*
> ```XML
>     shrinkwrapDependency = shrinkwrapDependency || shrinkwrap.dependencies[dependency];
> 
>     return shrinkwrapDependency && semver.satisfies(shrinkwrapDependency.version, version);
> ```

Do we just want it to satisfy, or do we want the selector to match exactly?

---

*[rush/rush/src/actions/GenerateAction.ts, line 263 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ80Gi7tJmz7rODzk:-KfEQ80Gi7tJmz7rODzl:b-qv15jt) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L263)):*
> ```XML
>       parameterLongName: '--force',
>       parameterShortName: '-f',
>       description: 'Forces cleaning of the node_modules folder before running "npm install".'
> ```

How is this different from `-c`?

---

*[rush/rush/src/actions/GenerateAction.ts, line 290 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQIWIwGIt7Fty8g3W:-KfEQIWIwGIt7Fty8g3X:bmj9ws5) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L290)):*
> ```XML
>     //    necessary dependencies. This will happen if someone is adding a new rush dependency,
>     //    or if someone is adding a dependency which already exists in another project
> ```

Get rid of this newline.

---

*[rush/rush/src/actions/GenerateAction.ts, line 295 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERbT7eYbnDPcq2lmQ:-KfERbT7eYbnDPcq2lmR:b-c1rpe9) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L295)):*
> ```XML
> 
>     // 4. Delete "common\node_modules"
>     GenerateAction._deleteCommonNodeModules(this._rushConfiguration,
> ```

This is weird. Can you get rid of the second parameter and just put the function call in an `if` statement?

---

*[rush/rush/src/actions/GenerateAction.ts, line 305 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERi-lUdKvgPzTesym:-KfERi-lUdKvgPzTesyn:b3kre) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L305)):*
> ```XML
>     // 6. Make sure the NPM tool is set up properly.  Usually "rush install" should have
>     //    already done this, but not if they just cloned the repo
>     console.log('');
> ```

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfES4j6LlhUov64W5XZ:binyfr7) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Use `require`.
</blockquote></details>

I was just talking with Pete about this. He thinks using require to load JSON is "wrong"

---

*[rush/rush/src/actions/GenerateAction.ts, line 203 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOJDrdvOpACdW-Ai0:-KfETAjNOCa5WbfTanqP:bgpxa8s) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L203)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Why are you prepending with `EOL`?
</blockquote></details>

Just to ensure that there is some spacing here, otherwise it is a bit jumbled and difficult to read

---

*[rush/rush/src/actions/GenerateAction.ts, line 208 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOdikIHk0Zu7IeoMp:-KfETJ4eEpCiumJgMQDM:bl7l8as) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L208)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Does lazy mode really serve much purpose anymore? It skips the fastest step.
</blockquote></details>

It doesn't skip the fastest step. It skips deleting the common folder. It also skips shrinkwrap, but we specifically do that to call attention to the fact that it has problems. Another way to think about this is: rush generate will choose to run either "--lazy" or "--clean" based on its needs. "--lazy" effectively says: "even though I think we should delete the node_modules, you are telling me not to. presumably to save some time."

---

*[rush/rush/src/actions/GenerateAction.ts, line 213 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOjh71K3j1ercG0FN:-KfEUTnSwPQho86VI7pY:bk3kh6f) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L213)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

I don't think you need to say it's changing "modes."
</blockquote></details>

updated text

---

*[rush/rush/src/actions/GenerateAction.ts, line 230 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEPv3Podsh6U0-w49a:-KfEUdQCCGVlgTG35bHb:bgquaki) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L230)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

I'd look in the root of the shrinkwrap first because that's where we prefer dependencies to live.
</blockquote></details>

Nope. It has to look under the package first. NPM will prefer to use the version under the project if it exists.

Example:

```
root
|
+--- foobar@3
|
+--- rush-test-package
       |
        +--- foobar@2
```

Now imagine that the user is asking to update test-package to use foobar@3. Right now they are going to be resolved to 2, which is wrong (implying we need to do a full rebuild). 

**Note** this comment made me find a bug, which is now fixed.

---

*[rush/rush/src/actions/GenerateAction.ts, line 238 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ21tgiN_srK0rF7F:-KfEWM_fFn0a1SP7KNTw:bbqpq15) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L238)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Do we just want it to satisfy, or do we want the selector to match exactly?
</blockquote></details>

It just needs to satisfy.

---

*[rush/rush/src/actions/GenerateAction.ts, line 263 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ80Gi7tJmz7rODzk:-KfESW31_CB189GOzmL2:b-7haelf) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L263)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

How is this different from `-c`?
</blockquote></details>

Perhaps it should be --clean, for consistency

---

*[rush/rush/src/actions/GenerateAction.ts, line 295 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERbT7eYbnDPcq2lmQ:-KfEXFPAGRtsyDKENu80:blvhbd7) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L295)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

This is weird. Can you get rid of the second parameter and just put the function call in an `if` statement?
</blockquote></details>

Agreed. I did a little refactoring of this.

---

*[rush/rush/src/actions/GenerateAction.ts, line 305 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERi-lUdKvgPzTesym:-KfEXIVCGi4cCGYGDZuM:b892xei) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L305)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

?
</blockquote></details>

It was already there :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[rush/rush/src/actions/GenerateAction.ts, line 186 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEORoDei151Cx4LpFp:-KfEZLZivqu2SofefQVg:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L186)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Instead of hitting the disk twice, log and return `true` if `require` below throws.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 199 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfENpXxPuM9MCcIB1bC:-KfEZK_rZSmOh0weCIs9:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L199)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

What do you mean "technically?
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 203 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOJDrdvOpACdW-Ai0:-KfEZJz6pqq6KbuqZ_xV:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L203)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

Just to ensure that there is some spacing here, otherwise it is a bit jumbled and difficult to read
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 208 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOdikIHk0Zu7IeoMp:-KfEZJJ9Qj4QUSXvdzHr:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L208)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

It doesn't skip the fastest step. It skips deleting the common folder. It also skips shrinkwrap, but we specifically do that to call attention to the fact that it has problems. Another way to think about this is: rush generate will choose to run either "--lazy" or "--clean" based on its needs. "--lazy" effectively says: "even though I think we should delete the node_modules, you are telling me not to. presumably to save some time."
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 213 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEOjh71K3j1ercG0FN:-KfEZIp9O8jcqBupA2cv:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L213)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

updated text
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 225 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEPic7ptAfyojpZnzb:-KfEZIMedrL1pcclXUyU:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L225)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

You call this `tempProjectName` above. Can you call them the same thing?
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 230 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEPv3Podsh6U0-w49a:-KfEZHt2wj2I6uKvKtwM:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L230)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

Nope. It has to look under the package first. NPM will prefer to use the version under the project if it exists.

Example:

```
root
|
+--- foobar@3
|
+--- rush-test-package
       |
        +--- foobar@2
```

Now imagine that the user is asking to update test-package to use foobar@3. Right now they are going to be resolved to 2, which is wrong (implying we need to do a full rebuild). 

**Note** this comment made me find a bug, which is now fixed.

</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 238 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ21tgiN_srK0rF7F:-KfEZHVlqFrOBI98n1T1:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L238)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

It just needs to satisfy.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 263 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQ80Gi7tJmz7rODzk:-KfEZHBO37SaJM6CzlpT:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L263)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

Perhaps it should be --clean, for consistency
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 290 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEQIWIwGIt7Fty8g3W:-KfEZFTV3Ltm9c4TT6vM:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L290)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Get rid of this newline.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 295 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERbT7eYbnDPcq2lmQ:-KfEZGJNxSo02K7gsjva:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L295)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

Agreed. I did a little refactoring of this.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 305 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfERi-lUdKvgPzTesym:-KfEZG_sn6IEGb0cbjN3:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L305)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

It was already there :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

iclanton commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json, line 1 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbKySlbEmivmAiNDU:-KfFbKyTU08lPhVq08Co:bh4k0ul) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json#L1)):*
> ```JSON
> {
> ```

Make sure you include the rush changes.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfFbtkE-EYrLeQPITgM:b-yb8epm) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

I was just talking with Pete about this. He thinks using require to load JSON is "wrong"
</blockquote></details>

It's not. That's what `require` is for.

---

*[rush/rush/src/actions/GenerateAction.ts, line 183 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbV_N3i64i4ynC9Lg:-KfFbV_N3i64i4ynC9Lh:b-nw1gnk) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/rush/rush/src/actions/GenerateAction.ts#L183)):*
> ```XML
>     try {
>       const shrinkwrapFile: string = path.join(rushConfiguration.commonFolder, 'npm-shrinkwrap.json');
>       shrinkwrap = JSON.parse(fsx.readFileSync(shrinkwrapFile).toString());
> ```

change this to `shrinkwrap = require(shrinkwrapFile);`.

---

*[rush/rush/src/actions/GenerateAction.ts, line 187 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbc-KbpzvPMkQYcrr:-KfFbc-KbpzvPMkQYcrs:b-4yr6dt) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/rush/rush/src/actions/GenerateAction.ts#L187)):*
> ```XML
>       /* no-op */
>     }
>     if (!shrinkwrap) {
> ```

newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json, line 1 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbKySlbEmivmAiNDU:-KfI-Mt2DhwA3mFwYLeD:bkb89yk) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json#L1)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Make sure you include the rush changes.
</blockquote></details>

I am going to do that manually. If I check in the Rush changefile now, the version bump will be applied but the package will not be published!

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfI-4bPd8jpyb0dBqF8:b-904utj) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

It's not. That's what `require` is for.
</blockquote></details>

His argument is that require is for "executables" and that JSON should be parsed using an actual json parser. Basically his point is we are calling "eval" on a file rather than "parsing the file as JSON". I will change it anyway, though.

---

*[rush/rush/src/actions/GenerateAction.ts, line 183 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbV_N3i64i4ynC9Lg:-KfHztvpFjdsHabuP8cw:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/rush/rush/src/actions/GenerateAction.ts#L183)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

change this to `shrinkwrap = require(shrinkwrapFile);`.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 187 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbc-KbpzvPMkQYcrr:-KfI-2lO5fBrtYEQJYRP:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/rush/rush/src/actions/GenerateAction.ts#L187)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json, line 1 at r2](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfFbKySlbEmivmAiNDU:-KfI-dMw5br8tAuqtaf-:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/c4cad0d65a30e8a5df8a55957c7f073067689764/common/changes/nickpape-generate-skip-work_2017-03-15-00-05.json#L1)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

I am going to do that manually. If I check in the Rush changefile now, the version bump will be applied but the package will not be published!
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfI-bf_YKFueVcQRj1r:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

His argument is that require is for "executables" and that JSON should be parsed using an actual json parser. Basically his point is we are calling "eval" on a file rather than "parsing the file as JSON". I will change it anyway, though.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

iclanton commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfIVTMxarOCuXRl0FxO:bnlt480) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, nickpape-msft (Nicholas Pape) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I buy that argument only in the context of a non-`require` parser doing something that the `require` parser doesn't do (like removing comments). Reading the file with `fs` and then calling `JSON.parse` is extra steps with no benefit and a tangible downside (namely - if the file is `require`d more than once, we'll hit the disk more than once).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

iclanton commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 7 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-:-KfIWC8uKfQMz9yUH_yx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nickpape-msft commented:
Reviewed 6 of 7 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[rush/rush/src/actions/GenerateAction.ts, line 180 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfELwTa5DvbpYu0PXvo-r1-180:-KfN4gsQVRz0jw6Y1URr:b32w0x) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L180)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Leave as a JSDoc comment?
</blockquote></details>

OK

---

*[rush/rush/src/actions/GenerateAction.ts, line 186 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfELwTa5DvbpYu0PXvo-r1-186:-KfN4aM-oeghUD3Kjy0M:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L186)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

I'd use `require` below and return if it throws.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfELwTa5DvbpYu0PXvo-r1-192:-KfN4a_Vis2XRvngXy7R:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

Just use `require`.
</blockquote></details>

Done.

---

*[rush/rush/src/actions/GenerateAction.ts, line 192 at r1](https://reviewable.io:443/reviews/microsoft/web-build-tools/133#-KfEONlBcrBKhuG5x6-D:-KfN4cTNrUGJ0utPk0Hw:b-896fix) ([raw file](https://github.com/microsoft/web-build-tools/blob/b20bc84a4517a4c9c9c1e475aa5d7ba46788c46e/rush/rush/src/actions/GenerateAction.ts#L192)):*
<details><summary><i>Previously, iclanton (Ian Clanton-Thuon) wrote…</i></summary><blockquote>

I buy that argument only in the context of a non-`require` parser doing something that the `require` parser doesn't do (like removing comments). Reading the file with `fs` and then calling `JSON.parse` is extra steps with no benefit and a tangible downside (namely - if the file is `require`d more than once, we'll hit the disk more than once).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/133)*
<!-- Sent from Reviewable.io -->

pgonzal commented on rush/rush/src/actions/GenerateAction.ts:
```diff
@@ -176,26 +277,67 @@ export default class GenerateAction extends CommandLineAction {
+      console.log(os.EOL + colors.bold('(Skipping "npm shrinkwrap")') + os.EOL);
+    } else {
+      GenerateAction._runNpmShrinkWrap(this._rushConfiguration);
+    }
```
We also realized that after these changes, the temp_modules don't need to be in Git at all any more.  We said the user interface should be like this:

- **rush install**:  Always regenerates the temp_modules before installing.  However, if shrinkwrap is out of date, then it fails (after zero seconds!), telling the person to run "rush upgrade".  (It's not automatic, because "upgrade" takes long and is a complicated PR approval.)

- **rush upgrade**:  Always does a full "generate" of everything, even if shrinkwrap is technically valid.  Supports the `--lazy` option, which is really a tool to help people who need to do multiple attempts as part of a full upgrade of everything.
